ocaml/xapi/xapi_host.ml

```diff
@@ -2923,6 +2923,76 @@ let emergency_reenable_tls_verification ~__context =
   Helpers.touch_file Constants.verify_certificates_path ;
   Db.Host.set_tls_verification_enabled ~__context ~self ~value:true
 
+(** Issue an alert if /proc/sys/kernel/tainted indicates particular kernel
+    errors. Will send only one alert per reboot *)
+let alert_if_kernel_broken =
+  (* This is an optimization. We only know for certain if the alert has been
+     sent if the toolstack hasn't been restarted. If it was restarted, we need
+     to check all the alerts since last reboot (and issue an alert for the
+     first time if needed *)
+  let which_alerts_have_been_sent =
+    ref
+      [
+        ((4, Api_messages.kernel_is_broken "MCE"), false)
+      ; (* processor reported a Machine Check Exception (MCE) *)
+        ((5, Api_messages.kernel_is_broken "BAD_PAGE"), false)
+      ; (* bad page referenced or some unexpected page flags *)
+        ((7, Api_messages.kernel_is_broken "BUG"), false)
+      ; (* kernel died recently, i.e. there was an OOPS or BUG *)
+        ((9, Api_messages.kernel_is_broken_warning "WARN"), false)
+      ; (* kernel issued warning *)
+        ((14, Api_messages.kernel_is_broken_warning "SOFT_LOCKUP"), false)
+        (* soft lockup occurred *)
+      ]
+  in
+  fun ~__context ->
+    (* Only add an alert if (a) we found a problem and
+       (b) an alert wasn't already issued for the currently booted kernel *)
+    which_alerts_have_been_sent :=
+      List.map
+        (fun (((alert_bit, alert_message) as key), has_alert_been_sent) ->
+          if not has_alert_been_sent then
+            let is_bit_tainted =
+              Unixext.string_of_file "/proc/sys/kernel/tainted" |> int_of_string
+            in
+            let is_bit_tainted = (is_bit_tainted lsr alert_bit) land 1 = 1 in
+            if is_bit_tainted then (
+              let self = Helpers.get_localhost ~__context in
+              let boot_time =
+                Db.Host.get_other_config ~__context ~self
+                |> List.assoc "boot_time"
+                |> float_of_string
+              in
+
+              let alert_already_issued_for_this_boot =
+                Helpers.call_api_functions ~__context (fun rpc session_id ->
+                    Client.Client.Message.get_all_records ~rpc ~session_id
+                    |> List.exists (fun (_, record) ->
+                           record.API.message_name = fst alert_message
+                           && API.Date.is_later
+                                ~than:(API.Date.of_unix_time boot_time)
+                                record.API.message_timestamp
+                       )
+                )
+              in
+
+              ( if not alert_already_issued_for_this_boot then
```

Comment:
I guess this is a fallback in case toolstak has restarted and the tracking of messages has failed already.

But in that case, what are we gaining here by doing the internal tracking? Just doing fewer calls? I'd rather drop the internal tracking and rely on the boot_time-based tracking, at the expense of making the alerts much stickier (they make the alert to reappear if the user dismisses them, while the kernel is tainted). It should make the code much more straightforward, especially since the get_all_records can be called just once.

Reply:
to make alerts sticky, we'd need to call get_all_records every time, not just once. otherwise, how would we know that the alert has been dismissed and needs to be issued again? we can do this instead, maybe:
* check messages on startup once
* use that to create internal tracking bitmap
* only use the internal tracking bitmap after that

Reply:
> to make alerts sticky, we'd need to call get_all_records every time, not just once.

I meant once per scheduled check, not just once. Note that now it's being called up to once per element in the list in every scheduled check.

> otherwise, how would we know that the alert has been dismissed and needs to be issued again?

I don't think there's a way, which is why the alerts would be sticky in the method I proposed.

> we can do this instead, maybe: [...]

That seems like it should work, yes. I recommend removing the elements from the list instead of using a boolean.